### PR TITLE
Optional memory profiler workflow

### PR DESCRIPTION
**Pull request by YigitElma**
This adds an optional profiling workflow which will profile the CPU memory usage of some memory-intensive routines (such as Free Boundary, Bounce Averages, Jacobian) and comment the results in a way similar to timing benchmarks. Also, this will plot the data for each function, as well as master plots and saves it to artifacts.

There are a couple of concerns. First, how reliable is this, which I think is good enough to see significant changes. Second, I just run them around 3 times, I think this is enough since memory profile is less volatile then timing.

### CODE REVIEW

**Comment by YigitElma** [comment]
> the memory benchmark took 6 hours to setup the tmate session... something weird going on there

My bad, I deleted a part of it since I don't want to use it, but there was one more step that I had to delete. It should work now...

File: tests/benchmarks/compare_mem_results.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+run_id = os.environ.get("GITHUB_RUN_ID")
+repo = os.environ.get("GITHUB_REPOSITORY")
+
+# GitHub artifact URL (raw links are not stable unless served via Pages)
+image_url = f"https://github.com/{repo}/actions/runs/{run_id}/artifacts"
+
+# ---------- commit message ----------
+msg = "### Memory benchmark result\n\n```diff\n"
+msg += (
+    "| {'Test Name':^38} | {'%Δ':^12} | {'Master (MB)':^18} | "
```
Comment (YigitElma) [inline tests/benchmarks/compare_mem_results.py:47]:
```suggestion
    f"| {'Test Name':^38} | {'%Δ':^12} | {'Master (MB)':^18} | "
```

File: tests/benchmarks/compare_mem_results.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    + "{'PR (MB)':^18} | {'Δ (MB)':^12} |\n"
```
Comment (YigitElma) [inline tests/benchmarks/compare_mem_results.py:48]:
```suggestion
    + f"{'PR (MB)':^18} | {'Δ (MB)':^12} |\n"
```

File: tests/benchmarks/compare_mem_results.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    + f"{'PR (MB)':^18} | {'Δ (MB)':^12} |\n"
+)
+msg += f"| {'-'*38} | {'-'*12} | {'-'*18} | {'-'*18} | {'-'*12} |\n"
+for i, name in enumerate(data_master.keys()):
+    peak_pr = data_pr[name]["mem"].max()
+    peak_ma = data_master[name]["mem"].max()
+    delta = peak_pr - peak_ma
+    sign = "+" if delta >= 0 else "-"
+    # only show color if the delta is significant
+    color = " " if abs((delta / peak_ma) * 100) < 7 else "-" if delta >= 0 else "+"
```
Comment (YigitElma) [inline tests/benchmarks/compare_mem_results.py:57]:
```suggestion
    color = " " if abs((delta / peak_ma) * 100) < 10 else "-" if delta >= 0 else "+"
```

**Comment by YigitElma** [comment]
Example plot if you download the artifact,
compare

File: tests/benchmarks/memory_funcs.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    QuasisymmetryTwoTerm,
+    get_equilibrium_objective,
+    get_fixed_boundary_constraints,
+    maybe_add_self_consistency,
+)
+from desc.optimize import LinearConstraintProjection, ProximalProjection
+
+
+@pytest.mark.skip()
+@pytest.mark.unit
```
Comment (f0uriest) [inline tests/benchmarks/memory_funcs.py:44]: probably better to just add a new mark for memory tests rather than marking them as unit + skip

File: tests/benchmarks/memory_funcs.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        ObjectiveFunction(
+            maybe_add_self_consistency(eq, get_fixed_boundary_constraints(eq)),
+        ),
+    )
+    objective.build(verbose=0)
+    x = objective.x(eq)
+
+    for _ in range(3):
+        objective.jac_scaled_error(x, objective.constants).block_until_ready()
+        gc.collect()
```
Comment (f0uriest) [inline tests/benchmarks/memory_funcs.py:60]: do we want this here? wouldn't this possibly reduce memory usage below what we'd see in practice?

File: tests/benchmarks/compare_mem_results.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    for i, name in enumerate(data_master.keys()):
+        peak_pr = data_pr[name]["mem"].max()
+        peak_ma = data_master[name]["mem"].max()
+        delta = peak_pr - peak_ma
+        # only show color if the delta is significant
+        color = " " if abs((delta / peak_ma) * 100) < 10 else "-" if delta >= 0 else "+"
+        percent_change = f"{(delta / peak_ma) * 100:.2f}" + " %"
+        delta = f"{delta:.2f}"
+        msg += (
+            f"{color} {name:<38} | {percent_change:^12} | {peak_ma:^18.1f} | "
```
Comment (f0uriest) [inline tests/benchmarks/compare_mem_results.py:54]: minor point but I prefer exponential notation `{blah:.3e}` to floating point

File: tests/benchmarks/memory_funcs.py (hunk 5 above)
Comment (YigitElma) [reply tests/benchmarks/memory_funcs.py]: Moved it to top with `jax.clear_caches()`

File: .github/workflows/jax_tests.yml [hunk 7]
```diff
@@ -2,12 +2,12 @@ name: Dependency test JAX
...
 on:
   pull_request:
-    types: [labeled]
+    types: [labeled, synchronize]
   workflow_dispatch:
 
 jobs:
   jax_tests:
-    if: ${{ github.event.label.name == 'test_jax' && github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch' }}
+    if: ${{ contains(github.event.pull_request.labels.*.name, 'test_jax') || github.event_name == 'workflow_dispatch' }}
```
Comment (YigitElma) [inline .github/workflows/jax_tests.yml:10]: We had to manually remove/add the label everytime we push a commit, this fixes that

**Comment by dpanici** [comment]
Make this required (maybe ake benchmarks not auto)

**Review by ddudt — DISMISSED** [review]
I agree with Rory's comment about scientific notion, but other wise looks good

**Review by dpanici — DISMISSED** [review]

**Review by dpanici — APPROVED** [review]

**Review by rahulgaur104 — APPROVED** [review]
